### PR TITLE
add support for gitlab ci

### PR DESCRIPTION
**Pull request by MaCXyLo**
missed gitlab ci support for the nixos generators project.
added as many formats as possible.

running example:
https://gitlab.com/MaCXyLo/nixos-generators/pipelines/99808542

### CODE REVIEW

File: nixos-generate [hunk 1]
```diff
@@ -35,6 +36,7 @@ Options:
 * --system: specify the target system (eg: x86_64-linux)
 * --cores : to control the maximum amount of parallelism. (see nix-build documentation)
 * -I KEY=VALUE: Add a key to the Nix expression search path.
+* --nixpkgs_ver: set the release version of nixpkgs
```
Comment (zimbatm) [inline nixos-generate]: I don't know if it's worth to have yet another shortcut. You get the same by using the `-I` argument. Eg: `-I nixpkgs=channel:nixos-19.09`

File: nixos-generate [hunk 2]
```diff
@@ -103,9 +109,14 @@ if [[ -z $format_path ]]; then
   abort "missing format. use --help for more details"
 fi
 
+if [[ -n $nixpkgs_ver ]]; then
+  args+=(--nixpkgs_ver "$nixpkgs_ver")
```
Comment (zimbatm) [inline nixos-generate]: `args` is now called `nix_args`

File: nixos-generate [hunk 3]
```diff
@@ -103,9 +109,14 @@ if [[ -z $format_path ]]; then
...
+fi
+
 nix_args+=(
   -I "nixos-config=$configuration"
   -I "format-config=$format_path"
+  -I "nixpkgs=https://github.com/NixOS/nixpkgs-channels/archive/nixos-$nixpkgs_ver.tar.gz"
```
Comment (zimbatm) [inline nixos-generate]: This would fail if `nixpkgs_ver` is unset.

File: nixos-generate (hunk 1 above)
Comment (rautends) [reply nixos-generate]: your right, wrote yesterday with lassulus on matrix about it.
a hint in the README.md is enough at that point, i think.

Comment (MaCXyLo) [reply nixos-generate]: your right at that point. will correct that.

**Comment by MaCXyLo** [comment]
> * +0 on the gitlab CI configuration. This is not going to be exercised on this repository and is bound to rot.

can you describe your objections in more detail, then can I adapt the gitlab ci file further?

File: nixos-generate (hunk 1 above)
Comment (zimbatm) [reply nixos-generate]: good idea

**Comment by zimbatm** [comment]
> can you describe your objections in more detail so i am able to further customize the gitlab ci script?
The only issue is that we don't have a Gitlab CI running against the repo right now. I am sure the gitlab-ci script is great and it seems like you put some effort into it.

/cc @adisbladis because he added a GitLab CI runner to our build host.

**Comment by adisbladis** [comment]
People are free to use whatever CI's they see fit for nix-community, and gitlab-ci does support Github repos too so that's surely an option.
It's up to the project maintainer to decide which one to use.

Currently we have buildkite and gitlab-ci set up, but I'd be willing to add more if required.

cc @Lassulus

**Comment by Lassulus** [comment]
I'm fine with having gitlab-ci support. I just don't know about building all images all the time though?

**Comment by zimbatm** [comment]
@adisbladis does it mean that the repo has to be mirrored somewhere on GitLab?

**Comment by MaCXyLo** [comment]
>[...] I just don't know about building all images all the time though?

continous integration ensures that all images can actually be built in your specific project.
so you are able to proof, if the actual nixpkgs release is faultless.

people can fork this repository, to build they're own images:
nobody needs nixos installed on they're host to build they're custom nixos image/container/vm,
they just can use the ci system.
you can easily disable ci jobs by adding a single dot at the beginning of the job name.
this ensures a wider spread of nixos over the internet.

**Comment by MaCXyLo** [comment]
> does it mean that the repo has to be mirrored somewhere on GitLab?

gitlab is able to mirror github repositories.
see: https://docs.gitlab.com/ee/user/project/repository/repository_mirroring.html

**Comment by adisbladis** [comment]
> @adisbladis does it mean that the repo has to be mirrored somewhere on GitLab?

Yes, though gitlab does that automatically and triggers on pull requests.

**Comment by zimbatm** [comment]
> gitlab is able to mirror github repositories if you are working on an open source project.
see: https://docs.gitlab.com/ee/user/project/repository/repository_mirroring.html

In that case I think we should also create a nix-community or on GitLab as well.

**Comment by zimbatm** [comment]
@MaCXyLo: GitLab CI should be active now. Can you try and push another commit on this branch?

**Comment by MaCXyLo** [comment]
@zimbatm 
first, you must bring this pull request to branch e.g. a sperate one or the master itself.
after that you can select the branch and trigger here the pipeline:
https://gitlab.com/nix-community/nixos-generators/pipelines

**Comment by zimbatm** [comment]
I think it's missing the webhooks from GitHub. Their github importer wasn't working so I had to point to the URL directly and the webhooks haven't been setup yet.

**Comment by rautends** [comment]
i tried to trigger the webhook.

**Comment by MaCXyLo** [comment]
i tried to trigger the webhook.

**Comment by zimbatm** [comment]
sorry this needs to be fixed first. Maybe @adisbladis can help out as I haven't managed to use the Import from GitHub feature properly in GitLab.

**Comment by adisbladis** [comment]
Gitlab CI should work now.

**Comment by MaCXyLo** [comment]
my pull request is available as a branch, but gitlab doesn't recognize the .gitlab-ci.yml file.
"This project is not currently set up to run pipelines." :/

don't have permissions to the nix-community repository.
@zimbatm @adisbladis 
can you take a look?

**Comment by MaCXyLo** [comment]
@zimbatm @adisbladis 
ping.

**Comment by zimbatm** [comment]
let's fix CI later

**Pull request merged**